src/wp-includes/blocks.php

```diff
@@ -1002,6 +1002,115 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
 	return '';
 }
 
+/**
+ * Runs the hooked blocks algorithm on the given content.
+ *
+ * This function is meant for internal use only.
+ *
+ * @since 6.6.0
+ * @access private
+ *
+ * @param string $content Serialized content.
+ * @param WP_Block_Template|WP_Post|array $context       A block template, template part, `wp_navigation` post object,
+ *                                                       or pattern that the blocks belong to.
+ * @return string The serialized markup with the ignored hooked blocks metadata applied.
+ */
+function apply_hooked_blocks_to_content( $content, $context, $callback = 'insert_hooked_blocks' ) {
+	$hooked_blocks = get_hooked_blocks();
+	if ( empty( $hooked_blocks ) && ! has_filter( 'hooked_block_types' ) ) {
+		return $content;
+	}
+
+	$blocks = parse_blocks( $content );
+
+	$before_block_visitor = make_before_block_visitor( $hooked_blocks, $context, $callback );
+	$after_block_visitor  = make_after_block_visitor( $hooked_blocks, $context, $callback );
+
+	return traverse_and_serialize_blocks( $blocks, $before_block_visitor, $after_block_visitor );
+}
+
+/**
+ * Accepts the serialized markup of a block and its inner blocks, and returns serialized markup of the inner blocks.
+ *
+ * @since 6.6.0
+ * @access private
+ *
+ * @param string $serialized_block The serialized markup of a block and its inner blocks.
+ * @return string
+ */
+function remove_serialized_parent_block( $serialized_block ) {
+	$start = strpos( $serialized_block, '-->' ) + strlen( '-->' );
+	$end   = strrpos( $serialized_block, '<!--' );
+	return substr( $serialized_block, $start, $end - $start );
+}
+
+/**
+ * Updates the wp_postmeta with the list of ignored hooked blocks where the inner blocks are stored as post content.
+ * Currently only supports `wp_navigation` post types.
+ *
+ * @since 6.6.0
+ * @access private
+ *
+ * @param stdClass $post Post object.
+ * @return stdClass The updated post object.
+ */
+function update_ignored_hooked_blocks_postmeta( $post ) {
+	/*
+	 * In this scenario the user has likely tried to create a navigation via the REST API.
+	 * In which case we won't have a post ID to work with and store meta against.
+	 */
+	if ( empty( $post->ID ) ) {
+		return $post;
+	}
+
+	/**
+	 * Skip meta generation when consumers intentionally update specific Navigation fields and omit the content update.
+	 * Or if the incoming post is not a Navigation post type.
+	 */
+	if ( ! isset( $post->post_content ) || ( isset( $post->post_type ) && 'wp_navigation' !== $post->post_type ) ) {
```

Comment:
Probably makes sense to combine the `post_content` and the `ID` check above it into a single statement.

Reply:
Or even put it separately? I don't hate how [we'd organized it in GB](https://github.com/WordPress/gutenberg/blob/6108134aae75d1bd4826256490c609cb29044cd8/packages/block-library/src/navigation/index.php#L1572-L1578), as each case is fairly different in nature, and having separate comments kinda makes sense IMO 🤔

Reply:
https://github.com/WordPress/wordpress-develop/pull/6604/commits/7878d8a84f48ce9c9eff873611575169d7efdeb8